### ISSUE TITLE
Unable to trigger onbeforeunload using ViewTransitions

### ISSUE THREAD

**Issue by sam-high** [withastro/astro#10309]
### Astro Info

```block
Astro                    v4.4.9
Node                     v18.18.0
System                   Linux (x64)
Package Manager          unknown
Output                   static
Adapter                  none
Integrations             none
```

### Describe the Bug

window.onbeforeunload functions can be used to prompt a user to confirm whether they want to leave a page or refresh a page that has unsaved changes. This functionality works as expected with Astro full-page navigation but, when using client-side routing with View Transitions the prompt is only shown when refreshing a page, no prompt is shown when navigating back or to a new page.

On my application users are accidently navigating back by clicking their mobile back button too many times. With no prompt, they are navigated away and lose their data.

Are there any known workarounds to prompt a user before the page changes using Astro client-side routing?

### What's the expected result?

With View Transitions enabled and a page with window.onbeforeunload set:

When a user clicks the back button, navigates to a new page, or refreshes their page, prompt that user to confirm whether they are sure they want to navigate away from their current page.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/withastro-astro-68qtzw?file=src%2Fpages%2Fpage2.astro

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by Fryuni** [comment]
Hey, thanks for opening an issue. There are a few things to consider here:

First, to stop the transition from happening you can override the `swap` function assigned to the `astro:before-swap` event. That is how some VT extensions add extra functionality, wrapping that function into another function:

```ts
document.addEventListener("astro:before-swap", (ev) => {
  // Confirm whether the user wants to navigate away from the page
  const answer = confirm("Are you sure you want to leave this page?");

  if (!answer) {
    // Replace swapping the page to a no-op;
    ev.swap = () => {};
  }
});
```

It might be a nice improvement to detect calls to `preventDefault` on that event and not do any swapping. Something to look up or bring up on the [Astro core repo](https://github.com/withastro/astro).

Second, that still won't work exactly as probably want. When you are not using VT `window.onbeforeunload` is specific to the current page, if you navigate to another page the listener is not there. But when using VT everything is happening in the _same context_. So if you add a listener in one page, that listener continues to run when the same event fires in another page because it is not unloaded.
Similarly, anything happening at the root of your script won't execute again after transitioning back and forth between the same pages.

So you can't rely on the current page being the page that defined the script just because the script is running. With the example above the alert will show on every page transition after you visit the page that registers it.
This is both a problem and an advantage. It means you don't necessarily actually lose any unsaved progress by navigating away, it will all depend on what you are using to manage the client-side state, depending on the library you are using you might not need that warning at all when transitioning only when closing the tab.

If you do want that warning when transitioning you need to make sure that you yourself clean the state before returning from the event handler so transitions on other pages don't trigger the unsaved work alert.

**Comment by Fryuni** [comment]
> It might be a nice improvement to detect calls to `preventDefault` on that event and not do any swapping.

What do you think @martrapp? We support preventing the default on the `before-preparation` event to turn the transition into a full browser navigation, should we support preventing the default on `before-swap` to abort the transition?

**Comment by Fryuni** [comment]
For reference:

[This line](https://github.com/withastro/astro/blob/718eed704ad1864cdedf0704d5020af1439671cd/packages/astro/src/transitions/events.ts#L182) would be conditioned on the default not being prevented.

[Here](https://github.com/withastro/astro/blob/718eed704ad1864cdedf0704d5020af1439671cd/packages/astro/src/transitions/router.ts#L451-L454) is where the handling for a prevented default of the `before-preparation` happen.

**Comment by sam-high** [comment]
> ```ts
> document.addEventListener("astro:before-swap", (ev) => {
>   // Confirm whether the user wants to navigate away from the page
>   const answer = confirm("Are you sure you want to leave this page?");
> 
>   if (!answer) {
>     // Replace swapping the page to a no-op;
>     ev.swap = () => {};
>   }
> });
> ```

Thank you for the speedy response. This is very helpful!

Your solution almost works perfectly,  the page data is no longer changing which is amazing but the URL changes to the destination URL. e.g. Showing /page2 data but the URL is now /

In my case, the state of the page is being intentionally reset when back navigation occurs. This state can also help me define when and when not to trigger the confirmation prompt so thankfully the message isn't shown on every screen afterwards.

**Comment by martrapp** [comment]
Hi @sam-high,
> but the URL changes to the destination URL. e.g. Showing /page2 data but the URL is now /

In addition to making swap a no-op, you should also set the target URL to where you came from.
But sadly, that is not sufficient to block a history back navigation. 
The reason for this is that on `history.back()` navigation the browser already switched the history and we must compensate for that. This would be the necessary extension:
```
if (!answer) {
  // Replace swapping the page to a no-op;
  ev.swap = () => {};
  
  ev.to = ev.from; // ok for "normal" forard navigation
  // getting history navigation right is more complex as the history has changed before we get notified
  if (ev.navigationType === "traverse" && ev.direction === "back") {
    history.pushState({scrollX, scrollY}, "", ev.from.href);
  }
}
```
Even though this might be a technically working solution, it feels wrong to do the whole preparation including loading the target's DOM just to skip the swap later on.

As @Fryuni pointed out, before astro:before-preparation` reacts to setting `preventDefault()` and I think this should be the way to go. When you currently opt out of preparation, Astro's view transition falls back to a full page load. Maybe we should add an option on the event where you can specify that `preventDefault()` should just cancel the default behavior of navigation and stay on the page.

Last thought: Currently, the `astro:before-swap` event is not cancelable. I wouldn't change that without need as there are already mechanisms to cancle the visual transition by calling `skipTransition()` and to re-define the swap() as shown in the example above.

**Comment by martrapp** [comment]
@sam-high & @Fryuni: My proposal would be to add a new attribute to the `astro:before-preparation` event: `whenCanceled: 'load' | 'ignore'` to state what should happen after `ev.preventDefault()` was called. Here 'load' is the default and selects the current behavior of delegating to the browser for a full page load. The 'ignore' value ignores and ends the initial navigation request. 

thoughts on this?

**Comment by Fryuni** [comment]
I'd prefer to keep the `.preventDefault()` to _prevent the default_ behavior of the ViewTransitions by switching to an alternative behavior and have a separate method, maybe `.abortTransition()` that would stop everything and revert everything to what it was before starting.

**Comment by Fryuni** [comment]
But that might not be very idiomatic web development. I don't know. I very rarely do browser-side logic.

**Comment by martrapp** [comment]
Very good suggestion! 

The dilemma is a bit that `e.preventDefault()` should cancel the default action associated with the event. As an astro view transition event, it has canceled view transitions so far. As a navigation event, it could also cancel navigation, which would be more similar to the semantics of the 'click' handler.  I have to admit that canceling navigation sounds more expectable, but that would be a breaking change from what we have now.

Maybe adding an `abortNavigation()` function is less irritating than programming the semantics of `preventDefault()` with the help of an additional attribute ;-)

**Comment by sam-high** [comment]
The idea of a function in `astro:before-preparation` or `astro:before-swap` sounds intuitive to me personally. 

I wanted to share some initial assumptions I had with this feature in the hopes they may help.

My initial assumption was that adding the below code would trigger the same prompt behaviour on both full-page and client-side navigation.
```
// Pop up browser confirmation
window.onbeforeunload = function () {
 return "Your work will be lost.";
};
```
The benefit I see with solving this approach is that it quickly enables built-in browser behaviour to prompt users with very little logic required.

My second assumption was that I could cancel or abort the client-side navigation somewhere like `astro:before-swap`
and add in my own logic, using `confirm()` or a custom modal to prompt the user.

If a function such as `abortNavigation()` or `abortTransition()` was documented on the https://docs.astro.build/en/guides/view-transitions page perhaps with an example under `astro:before-preparation` or `astro:before-swap` (wherever you choose to implement it), that would feel intuitive to use for me.

The benefit I see to this approach is that enables greater control and customization over the navigation process while expanding the useful features your client-side navigation provides. Having this feature in your docs would also make this the obvious solution moving forward.

**Comment by matthewp** [comment]
We don't fire `beforeunload` because the document isn't unloaded. We don't want to simulate real browser events because they are used for things that need to happen when the window closes.

**Comment by martrapp** [comment]
TL;DR: Hey @sam-high, I would happily help you with getting the code above robust for your particular use case. But I'm afraid, there is no general solution for this issue to be build into Astro's core.

  ---

The solution sketched [above](https://github.com/withastro/astro/issues/10309#issuecomment-1976187141) has several weaknesses. 
- As history traversals (popstate events) can not be canceled, they must be compensated after they happen. 
- Currently, that fragment only attempts to compensate for backward history traversals, but must also consider compensation for forward history traversals.
- By pushing new state causes all history entries "to the right" to be lost. A viable solution should use `history.go()` to traverse existing history entries without destroying them.
- Using `go()` triggers more `popstate` events and carries a high risk of looping.

All these points can be tackled, but there doesn't seem to be a general solution for the next point:
Users may use the browsers history menus to do wide jumps (more than one page) backwards and forwards through history. Compensation should take that distance into account, i.e. calling `go(3)` or `go(-4)`.
 
I have no glue how to figure out that distance in the general case.
- We can not use the index in Astro's history state, as it is not guaranteed to be monotone, i.e. it might go `0, 1, 2, 0, 1` when some transitions have a `data-astro-reload` attribute or it might even go `0, 1, -, -, 0, 1` when crossing pages without the `<ViewTransitions>` component.
- We can not use information from the Navigation API as this is not widely available.

**Comment by sam-high** [comment]
Sorry to hear there isn't an obvious way to create a general solution.

With the code you provided and some minor hacks, I have managed to come up with a solution for my use case.

I appreciate all of your help very much!

```
if (!answer) {
  // Replace swapping the page to a no-op;
  ev.swap = () => {};
  
  ev.to = ev.from; // ok for "normal" forard navigation
  // getting history navigation right is more complex as the history has changed before we get notified
  if (ev.navigationType === "traverse" && ev.direction === "back") {
    history.pushState(
      { index: (history.state?.index ?? 0) + 1, scrollX, scrollY }, 
      "", 
      ev.from.href
    );  
  }
}
```
